### DISCUSSION TITLE
STM32 Hash Peripheral

### DISCUSSION THREAD

**Discussion by mbanducci-sandc** [zephyrproject-rtos/zephyr#71407] [Q&A] [answered]
Hi all,

I'm working on a system based on an STM32H755 and I'm having some trouble getting the HASH peripheral going. I did some digging, and it appears that Zephyr combines ciphers and hashes under the `crypto` driver, but ST splits the two into unique peripherals. I did find ST HAL code for the HASH peripheral, but it doesn't appear to make it into `crypto_stm32.c`.

Is there a piece I'm missing with regard to the STM32 HASH driver, or is this something that could use work to bring it to full functionality?

Thank you!
Mike

**Comment by ajarmouni-st** [comment, marked as answer]
Hi, `crypto_stm32.c` does not yet implement Zephyr's `crypto_driver_api` Hash functions (`hash_begin_session`, `hash_free_session`, `hash_async_callback_set`). So for now you will have to either use STM32 HAL directly or implement the Hash functions in `crypto_stm32.c` (an example of a driver that implements them: `drivers/crypto/crypto_intel_sha.c`) & then use Zephyr Crypto API.

**Reply by mbanducci-sandc** [reply]
Understood, thank you!
